### DISCUSSION TITLE
Unable to generate a extxyz database

### DISCUSSION THREAD

**Discussion by jmguevarav** [mir-group/nequip#334] [Q&A] [answered]
I am new using nequip. I was able to run the examples without problem, but when I try to use my own database I found the following error:

> Torch device: cpu
Traceback (most recent call last):
  File "/home/jmgv/miniconda3/envs/nequip/lib/python3.10/site-packages/nequip/utils/auto_init.py", line 232, in instantiate
    instance = builder(**positional_args, **final_optional_args)
  File "/home/jmgv/miniconda3/envs/nequip/lib/python3.10/site-packages/nequip/data/dataset.py", line 873, in __init__
    self.ase_args.update(ase_args)
TypeError: 'NoneType' object is not iterable

> The above exception was the direct cause of the following exception:

> Traceback (most recent call last):
  File "/home/jmgv/miniconda3/envs/nequip/bin/nequip-train", line 8, in <module>
    sys.exit(main())
  File "/home/jmgv/miniconda3/envs/nequip/lib/python3.10/site-packages/nequip/scripts/train.py", line 72, in main
    trainer = fresh_start(config)
  File "/home/jmgv/miniconda3/envs/nequip/lib/python3.10/site-packages/nequip/scripts/train.py", line 148, in fresh_start
    dataset = dataset_from_config(config, prefix="dataset")
  File "/home/jmgv/miniconda3/envs/nequip/lib/python3.10/site-packages/nequip/data/_build.py", line 78, in dataset_from_config
    instance, _ = instantiate(
  File "/home/jmgv/miniconda3/envs/nequip/lib/python3.10/site-packages/nequip/utils/auto_init.py", line 234, in instantiate
    raise RuntimeError(
RuntimeError: Failed to build object with prefix `dataset` using builder `ASEDataset`

I do not understand the source of this. The only lines that I changed from example.yaml are the following:

```
dataset: ase
dataset_file_name: platinum_clusters.extxyz
ase_args:
format: extxyz
include_keys:
  - energy
key_mapping:
  energy: total_energy
chemical_symbol_to_type:
  Pt: 0
```

Finally, my toy database was generated using ASE and looks like this:
```
17
Properties=species:S:1:pos:R:3:forces:R:3:energies:R:1 energy=14.277781233330693 user_label=1.0 free_energy=14.277781233330693 pbc="F F F"
Pt      -2.95723893      -2.66386844       0.48978474       0.72144578       0.74238004      -0.52204204       1.14769647
Pt       3.50053273      -0.99838087      -1.31969386      -0.55703873       0.89134386       1.15354519       0.94975695
Pt      -0.45888828      -2.05778392       0.77322413      -0.03998858      -0.02732856      -0.15430488       0.61574551
Pt       1.04201759      -1.60733438      -1.68165195       0.37003520      -0.16704142       0.69070160       0.79275206
Pt       4.09873035       0.24355516       0.90991805      -1.46281733      -0.37216940      -1.25808534       1.05635175
Pt      -1.41666408      -2.23541882      -1.62142297      -0.37573879      -0.30145019       0.52059060       0.88583652
Pt       2.13104544      -1.57231569       0.87327081       0.34799713       0.20860768      -0.37301507       0.63068709
Pt       1.49151195       0.88850974       1.08310282       0.10626615       0.11920175       0.15196518       0.51310081
Pt      -2.05689205       0.20559383      -1.52550671      -0.75518681       0.26030454       0.19191817       0.75928777
Pt      -3.48527335       2.13918353      -0.57236220       0.60524387      -0.53235210       0.43475408       1.11083408
Pt       0.69126106       3.02632282      -0.23261989      -0.02888316      -0.41328181       0.13998210       0.92285656
Pt      -1.11014408       0.41781405       1.12710010       0.08118040      -0.00834521      -0.26759353       0.50815406
Pt      -1.59347447       2.88263163       0.87669274      -0.01337742      -0.09841018      -0.54336427       1.04941986
Pt       0.35661845       0.82144764      -1.49924003       0.67145781       0.34404388       0.50087853       0.67145993
Pt       2.79960580       1.57081634      -1.01537369       0.28784764      -0.66735395       0.36790846       0.77225595
Pt       0.51752353      -0.83700541       2.84313144       0.05892466       0.12145161      -1.04580885       1.00910326
Pt      -3.55027165      -0.22376722       0.49164646      -0.01736782      -0.09960054       0.01197007       0.88248261

```

**Comment by Linux-cpp-lisp** [comment, marked as answer]
I think you're missing indentation on the `format` line after `ase_args:`.  Like in Python, indentation is meaningful in YAML and determines subkeys.

**Comment by jmguevarav** [comment]
That was it. Now it works like a charm. 

Thank you very much!

**Reply by Linux-cpp-lisp** [reply]
Glad to hear!

**Reply by DQM520** [reply]
> That was it. Now it works like a charm.
> 
> Thank you very much!
May I ask how do you get the formal extxyz file using ASE? I am new to nequip and I use lammps dump command 1 all custom ${DUMP_FREQ} methane.lammpstrj id type x y z to get the xyz file. I don't know how to convert it into extxyz file contains force and energy. Could you give me some advice? My questions attached https://github.com/mir-group/nequip/discussions/463

**Comment by KellekKan** [comment]
I am also new in using Nequip,  I was able to run the examples without problem, but when I try to use the Uracil database on the sDGML website, by firstly changing the databaset_url to #dataset_url: http://quantum-machine.org/gdml/data/npz/md17_uracil.npz, changing what are necessarily to be changed in the example,yaml file for the npz dataset, and secondly aiming to use the extxyz format for the training process. Bot h methods resulted in errors, for the first one (npz), it is not possible to download the npz format, and for the second one (extxyz) after downing the extxyz format from the website, i removed all other data points but just left the first point and added "Properties=species:S:1:pos:R:3:forces:R:3:energies:R:1 energy=-260120.1104989303", then after hitting nequip-train uracil.extxyz, it resulted in such errors (Attached). 
![image](https://github.com/mir-group/nequip/assets/119727053/745dbfc6-8b48-484f-9f3a-55ddb14a9ca5)


Thank you for your time.

**Reply by Linux-cpp-lisp** [reply]
Hi @KellekKan

This looks like a system issue, and has nothing to do with the dataset itself. You can try setting `NEQUIP_NUM_TASKS=1` to prevent `nequip` from multithreading in the dataset processing, since your system seems to have severe process limits.

**Reply by KellekKan** [reply]
Thank you for the prompt reply, may I ask for the NEQUIP_NUM_TASKS=1 command, where can I set it or do I just need to input in the command line ?  
Besides, could you please give me some instructions if I am wrong for setting the .extxyz file and the example.yaml ? 
For the extxyz settings, I wrote it in this way, is it correct ?
![image](https://github.com/mir-group/nequip/assets/119727053/e590405a-e3c4-4623-b149-37b050055516)

For the settings in example.yaml, do i need the option for the npz_fixed_field_keys (indicated by the yellow arrow), while I am using the extxyz format, and for the chemical_symbols (red arrow), do I need to put H:0, C:1 and so on? 
![image](https://github.com/mir-group/nequip/assets/119727053/4759cb1f-c991-4d54-8cec-4bbd5d042a64)

After going through the instructions, I am still confused for the use of user_label=1.0, could you also please explain me ?


Thanks again

**Reply by Linux-cpp-lisp** [reply]
`NEQUIP_NUM_TASKS=1` is an environment variable, i.e.
```
NEQUIP_NUM_TASKS=1 nequip-train ...
```


You can get rid of `npz_*` keys and also `key_mapping` in this case I think, all those field names are standard and should be handled by default. You can ignore `user_label`, that's an example of what to do if you have custom nonstandard fields in your data.

**Reply by KellekKan** [reply]
Thanks for the prompt reply, I encounter this issue, 
![image](https://github.com/mir-group/nequip/assets/119727053/768f482b-7edb-491a-a064-497c4d649538)
Besides, I am also wondering why the programme works nicely for the example system, no matter for the example or full, as well as the following evaluation and deploy processes. But when I changed the entire system to malonldehyde, which is also the npz dataset downloaded from the website in the example. The programme does not work anymore. 

Indicated that I just changed the url for downloading the data, added and O in the chemcial_symbols section, and certain filing names have been changed. 

I then obtained the same error massage as shown above. 

Besides, will the be programme settings be changed permanently when I used this command  NEQUIP_NUM_TASKS=1 ?

**Reply by Linux-cpp-lisp** [reply]
> Besides, will the be programme settings be changed permanently when I used this command NEQUIP_NUM_TASKS=1 ?

No, this is a per-run environment variable.

This is a bizarre error; my assumption from the `datetime` error is that something is just corrupted in your filesystem... can you load the numpy file just in a plain Python REPL without `nequip`? (It's just a plain `npz` file.)

Otherwise it looks like your install is somehow corrupt and you may just need to rebuild your conda env from scratch.

**Reply by KellekKan** [reply]
Regarding to this issue, if one uses .npz and would like to download them from the website used in the full.yaml example, the downloading process will not be processed without any error message, but if I change the dataset_url keyword to .zip from the same website, the downloading process works well sometime and the training will be started soon.

It seems that my installation is correct, and my internet connection is fine as well.

**Reply by Linux-cpp-lisp** [reply]
Just to clarify, do you have issues when you use an extXYZ dataset?  Both the downloading feature and the NPZ dataset format are largely for quick benchmarking, and are not as thoroughly tested in production as the extXYZ path.
